### ISSUE TITLE
errorMessagesForSchema returns wrong data

### ISSUE THREAD

**Issue by phifa** [SeasonedSoftware/domain-functions#55]
```ts
  const securityUpdateAction = useActionData<typeof action>() as any;
  if (securityUpdateAction) {
    const errorMessagesFormattedByDf = errorMessagesForSchema(
      securityUpdateAction.inputErrors,
      Schema
    );
  }

```

The `securityUpdateAction` returns `inputErrors` with this shape: 

```ts
[
    {
        "path": [
            "newPassword"
        ],
        "message": "String must contain at least 8 character(s)"
    },
    {
        "path": [
            "newPasswordRepeat"
        ],
        "message": "String must contain at least 8 character(s)"
    },
    {
        "path": [
            "newPassword",
            "newPasswordRepeat"
        ],
        "message": "Passwords don't match"
    }
]
```

The `Schema` looks like this: 

```ts
const Schema = z
  .object({
    oldPassword: z.string(),
    newPassword: z.string().min(8),
    newPasswordRepeat: z.string().min(8),
  })
  .refine((data) => data.newPassword === data.newPasswordRepeat, {
    message: "Passwords don't match",
    path: ["newPassword", "newPasswordRepeat"], // path of error
  })
```

The output after parsing looks like this in the browser console: 

<img width="380" alt="Bildschirmfoto 2022-12-12 um 20 01 38" src="https://user-images.githubusercontent.com/3061129/207142683-7eba639b-e377-4fc0-9fd1-ff6c8479bb4a.png">

In the docs it shows the shape what is to be expected: 
https://github.com/SeasonedSoftware/domain-functions#errormessagesforschema

```ts
errorForSchema(result.inputErrors, schema)
/*
{
  email: ['Must not be empty', 'Must be a string'],
  password: ['Must not be empty']
}
*/
```
Basically: `Record<string,string[]>`

**Comment by diogob** [comment]
I just found out that the same inputs produce different results in the browser and on deno. Will look into it so we can write an implementation that will work on all platforms.

**Comment by diogob** [comment]
@phifa regardless of the bug that is messing with the shape of the return of `errorMessagesForSchema`, your `path`  for the refinement should be the path for one field in the schema. Check [this example](https://zod.dev/?id=customize-error-path) in the Zod documentation and you will see that it points to the `confirm` field only.

Even if I fix the bug, that path would still be ignored as it does not exist in the schema.

When you write the path `["newPassword", "newPasswordRepeat"]` it points to an object with the shape: `{ newPassword: { newPasswordRepeat: ... } }`

**Comment by diogob** [comment]
Upon further investigation I understood that the code is indeed working as expected, and the only reason **it seems** to behave differently in the deno console is because deno does not print the result in the same way the browser.

The resulting shape is a record of arrays, the issue is that the code allows nesting inside the string object, which creates a very weird behaviour. The string inside the array behaves both as a string and as an object.

I'll mark this as a won't fix since the results are consistent across platforms, and although the result is not perfectly captured in the type at least we can trust the type-system (since I can still manipulate the result as a `Record<string,string[]>`). In other words, our runtime result is a super-type that will happen in a few edge-cases.

For more details on how to reproduce my reasoning see the example bellow.

To illustrate I have used a minimal setup using deno version 1.28.3:
```ts
import { z } from 'https://deno.land/x/zod@v3.19.1/mod.ts'
import type {
  ErrorWithMessage,
  SchemaError,
  ErrorResult,
  ErrorData,
  AtLeastOne,
} from './types.ts'

const errors = [
    {
        "path": [
            "newPassword"
        ],
        "message": "String must contain at least 8 character(s)"
    },
    {
        "path": [
            "newPasswordRepeat"
        ],
        "message": "String must contain at least 8 character(s)"
    },
    {
        "path": [
            "newPassword",
            "newPasswordRepeat"
        ],
        "message": "Passwords don't match"
    }
]

const schema = z
  .object({
    oldPassword: z.string(),
    newPassword: z.string().min(8),
    newPasswordRepeat: z.string().min(8),
  })
  .refine((data) => data.newPassword === data.newPasswordRepeat, {
    message: "Passwords don't match",
    path: ["newPassword", "newPasswordRepeat"], // path of error
  })

type NestedErrors<SchemaType> = {
  [Property in keyof SchemaType]: string[] | NestedErrors<SchemaType[Property]>
}

function errorMessagesForSchema<T extends z.ZodTypeAny>(
  errors: SchemaError[],
  _schema: T,
): NestedErrors<z.infer<T>> {
  type SchemaType = z.infer<T>
  type ErrorObject = { path: string[]; messages: string[] }

  const nest = (
    { path, messages }: ErrorObject,
    root: Record<string, unknown>,
  ) => {
    const [head, ...tail] = path
    root[head] =
      tail.length === 0
        ? messages
        : nest(
            { path: tail, messages },
            (root[head] as Record<string, unknown>) ?? {},
          )
    return root
  }

  const compareStringArrays = (a: string[]) => (b: string[]) =>
    JSON.stringify(a) === JSON.stringify(b)

  const toErrorObject = (errors: SchemaError[]): ErrorObject[] =>
    errors.map(({ path, message }) => ({
      path,
      messages: [message],
    }))

  const unifyPaths = (errors: SchemaError[]) =>
    toErrorObject(errors).reduce((memo, error) => {
      const comparePath = compareStringArrays(error.path)
      const mergeErrorMessages = ({ path, messages }: ErrorObject) =>
        comparePath(path)
          ? { path, messages: [...messages, ...error.messages] }
          : { path, messages }
      const existingPath = memo.find(({ path }) => comparePath(path))

      return existingPath ? memo.map(mergeErrorMessages) : [...memo, error]
    }, [] as ErrorObject[])

  const errorTree = unifyPaths(errors).reduce((memo, schemaError) => {
    const errorBranch = nest(schemaError, memo)

    return { ...memo, ...errorBranch }
  }, {}) as NestedErrors<SchemaType>

  return errorTree
}

class InputError extends Error {
  path: string

  constructor(message: string, path: string) {
    super(message)
    this.name = 'InputError'
    this.path = path
  }
}
```

And when we extract the errors this is what happens:

```ts

const e = errorMessagesForSchema(errors,schema)
undefined
> e
{
  newPassword: [ "String must contain at least 8 character(s)" ],
  newPasswordRepeat: [ "String must contain at least 8 character(s)" ]
}
> e["newPassword"]
[ "String must contain at least 8 character(s)" ]
> 
undefined
> e["newPassword"]["newPasswordRepeat"]
[ "Passwords don't match" ]
```

**Comment by diogob** [comment]
BTW @phifa I appreciate the detailed reporting, that made my investigation quite easy.

**Comment by phifa** [comment]
@diogob interesting, i had used the same schema before with zod's `flatten()` method, and it worked nicely. It added the refinement error for both fields, so in the UI I had the error displayed under both inputs. How can I achieve the same result with `errorMessagesForSchema()`? Do I have to use the `refine()` twice? changing the path for the second refinement?

**Comment by phifa** [comment]
i can confirm, this produces the wanted behaviour

```
export const UserSecuritySettingsUpdateData = z
  .object({
    oldPassword: z.string(),
    newPassword: z.string().min(8),
    newPasswordRepeat: z.string().min(8),
  })

  .refine((data) => data.newPassword === data.newPasswordRepeat, {
    message: "Passwords don't match",
    path: ["newPassword"], // path of error
  })
  .refine((data) => data.newPassword === data.newPasswordRepeat, {
    message: "Passwords don't match",
    path: ["newPasswordRepeat"], // path of error
  })
```
